presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotBrokerPageSourceBase.java

```diff
@@ -0,0 +1,384 @@
+    {
+        JsonNode numServersResponded = jsonBody.get("numServersResponded");
+        JsonNode numServersQueried = jsonBody.get("numServersQueried");
+
+        if (numServersQueried == null || numServersResponded == null || numServersQueried.asInt() > numServersResponded.asInt()) {
+            throw new PinotException(
+                PINOT_INSUFFICIENT_SERVER_RESPONSE,
+                Optional.of(pql),
+                String.format("Only %s out of %s servers responded for query %s", numServersResponded.asInt(), numServersQueried.asInt(), pql));
+        }
+
+        JsonNode exceptions = jsonBody.get("exceptions");
+        if (exceptions != null && exceptions.isArray() && exceptions.size() > 0) {
+            // Pinot is known to return exceptions with benign errorcodes like 200
+            // so we treat any exception as an error
+            throw new PinotException(
+                PINOT_EXCEPTION,
+                Optional.of(pql),
+                String.format("Query %s encountered exception %s", pql, exceptions.get(0)));
+        }
+    }
+
+    protected static String asText(JsonNode node)
+    {
+        checkState(node.isValueNode());
+        if (node.isArray()) {
+            String[] results = new String[node.size()];
+            for (int i = 0; i < node.size(); i++) {
+                results[i] = asText(node.get(i));
+            }
+            return Arrays.toString(results);
+        }
+        return node.isNull() ? null : node.asText();
+    }
+
+    @Override
+    public long getSystemMemoryUsage()
+    {
+        return 0;
+    }
+
+    @Override
+    public void close()
+    {
+        finished = true;
+    }
+
+    protected int issueQueryAndPopulate(
+            PinotQueryGenerator.GeneratedPinotQuery pinotQuery,
+            List<BlockBuilder> blockBuilders,
+            List<Type> types)
+    {
+        return doWithRetries(PinotSessionProperties.getPinotRetryCount(session), (retryNumber) -> {
+            String queryHost;
+            Optional<String> rpcService;
+            if (pinotConfig.getRestProxyUrl() != null) {
+                queryHost = pinotConfig.getRestProxyUrl();
+                rpcService = Optional.ofNullable(pinotConfig.getRestProxyServiceForQuery());
+            }
+            else {
+                queryHost = clusterInfoFetcher.getBrokerHost(pinotQuery.getTable());
+                rpcService = Optional.empty();
+            }
+            Request.Builder builder = Request.Builder
+                    .preparePost()
+                    .setUri(URI.create(String.format(getQueryUrlTemplate(), queryHost)));
+            String body = clusterInfoFetcher.doHttpActionWithHeaders(builder, Optional.of(String.format(getRequestPayloadTemplate(), pinotQuery.getSql())), rpcService);
+            return populateFromQueryResults(pinotQuery, blockBuilders, types, body);
+        });
+    }
+
+    abstract String getQueryUrlTemplate();
```

Comment:
Comments around the abstract methods to signify their purpose ?

Reply:
will do